core/trino-main/src/main/java/io/trino/cost/StatsCalculator.java

```diff
@@ -28,16 +28,18 @@
      * @param sourceStats The stats provider for any child nodes' stats, if needed to compute stats for the {@code node}
      * @param lookup Lookup to be used when resolving source nodes, allowing stats calculation to work within {@link IterativeOptimizer}
      * @param types The type provider for all symbols in the scope.
+     * @param tableStatsProvider The table stats provider.
      */
     PlanNodeStatsEstimate calculateStats(
             PlanNode node,
             StatsProvider sourceStats,
             Lookup lookup,
             Session session,
-            TypeProvider types);
+            TypeProvider types,
+            TableStatsProvider tableStatsProvider);
```

Comment:
Why does `TableStatsProvider` needs to be 1st class citizen in `StatsCalculator`. It seems it should be local to `TableScanStatsRule`

Reply:
> I am not following. @sopel39 what are you suggesting?

I suggest introducing some kind of context to `doCalculate` rather than adding another explicit parameter. I think @gaurav8297 has similar proposal for expression optimizer/analyzer results.

Reply:
> I suggest introducing some kind of context to `doCalculate` rather than adding another explicit parameter.

We still need to pass a parameter.

I agree we could introduce a more generic context as well.
I don't see a need for this here yet, can be a follow up